### ISSUE TITLE
0.7.33 rules bug

### ISSUE THREAD

**Issue by laweschan** [CloudFire-LLC/cloudfire#1929]
**Describe the bug**
whatever new build or migrate from 0.7.33, once update or change or add rule, container firezone will crashed.

**To Reproduce**
1. new build 
2. access success
3. update logo, update password, add device success.
4. once add rules, either deny or allow, the container will crash and not function immediately

**Expected behavior**
the firewall rules should work like 0.7.32 or before

**Screenshots / Logs**
Crash dump is being written to: erl_crash.dump...done
16:35:01.906 erl_level=info application=ecto_sql domain=elixir file=lib/ecto/migrator.ex function=log/2 line=707 mfa=Ecto.Migrator.log/2 module=Ecto.Mig                                                                    rator pid=<0.95.0> [info] Migrations already up
16:35:02.851 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/cowboy2_adapter.ex function=start_link/3 line=103 mfa=Phoenix.En                                                                    dpoint.Cowboy2Adapter.start_link/3 module=Phoenix.Endpoint.Cowboy2Adapter pid=<0.375.0> [info] Running FzHttpWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:1                                                                    3000 (http)
16:35:02.855 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/supervisor.ex function=log_access_url/2 line=425 mfa=Phoenix.End                                                                    point.Supervisor.log_access_url/2 module=Phoenix.Endpoint.Supervisor pid=<0.331.0> [info] Access FzHttpWeb.Endpoint at https://wgvpn.ecs-apps.com
16:35:03.026 erl_level=notice domain=otp file=application_controller.erl function=info_exited/3 line=2119 mfa=:application_controller.info_exited/3 modu                                                                    le=application_controller pid=<0.44.0> [notice] Application fz_wall exited: FzWall.Application.start(:normal, []) returned an error: shutdown: failed to                                                                     start child: FzWall.Server
    ** (EXIT) an exception was raised:
        ** (Protocol.UndefinedError) protocol String.Chars not implemented for {:ok, "445"} of type Tuple. This protocol is implemented for the followin                                                                    g type(s): Atom, BitString, CIDR, Date, DateTime, Decimal, FileSize.Bit, FileSize.Byte, Float, Floki.Selector, Floki.Selector.AttributeSelector, Floki.S                                                                    elector.Combinator, Floki.Selector.Functional, Floki.Selector.PseudoClass, FzHttp.Types.IPPort, Hex.Solver.Assignment, Hex.Solver.Constraints.Empty, Hex                                                                    .Solver.Constraints.Range, Hex.Solver.Constraints.Union, Hex.Solver.Incompatibility, Hex.Solver.PackageRange, Hex.Solver.Term, Integer, List, NaiveDateT                                                                    ime, Phoenix.LiveComponent.CID, Postgrex.Copy, Postgrex.INET, Postgrex.Query, RemoteIp.Block, Time, URI, Version, Version.Requirement
            (elixir 1.14.3) lib/string/chars.ex:3: String.Chars.impl_for!/1
            (elixir 1.14.3) lib/string/chars.ex:22: String.Chars.to_string/1
            (fz_wall 0.7.33) lib/fz_wall/cli/helpers/nft.ex:273: FzWall.CLI.Helpers.Nft.get_elem/3
            (fz_wall 0.7.33) lib/fz_wall/cli/helpers/nft.ex:40: FzWall.CLI.Helpers.Nft.add_elem/4
            (elixir 1.14.3) lib/enum.ex:980: anonymous fn/3 in Enum.each/2
            (elixir 1.14.3) lib/enum.ex:4307: anonymous fn/3 in Enum.each/2
            (elixir 1.14.3) lib/enum.ex:4751: Enumerable.List.reduce/3
            (elixir 1.14.3) lib/enum.ex:4307: Enum.each/2

**Platform (please complete the following information):**
 - Firezone 0.7.33
 - OS: Fedora
 - Distro: 38
 - Kernel version: 6.4. 11-100
 - Deployment method: podman

**Comment by toabi** [comment]
Got bit by this, upgraded from 0.7.32 to 0.7.33 and wanted to trie out rules for the first time and boom. Adding the Allow-Rule was in my case the issue. The deny rule worked.

Downgrade to 0.7.32, deleting the rules, upgrading again is a working fix to run on 0.7.33.
Or not having any rules.

**Comment by jamilbk** [comment]
Thanks for the report! Will have a fix out soon.

**Comment by wallarug** [comment]
Count me in for this error too!  Thanks.  Cian

**Comment by wallarug** [comment]
Hi @jamilbk,

Thanks for jumping on this issue.   Here is some details after I updated to 0.7.34.

Tried just re-running the bash script as per the docs.  It downloaded the latest firezone-firezone:0.7.34 container image for docker.  It still crashed in the same way.

I then wiped all the docker containers and tried from absolute scratch.  Same behaviour.  Crashing in the firezone-firezone-1 container every 10 seconds.

Here is the output from the docker log:

```
Crash dump is being written to: erl_crash.dump...done
Creating WireGuard interface wg-firezone
Adding route 100.64.0.0/10 for interface wg-firezone
Adding route fd00::/106 for interface wg-firezone
06:10:08.194 erl_level=info application=ecto_sql domain=elixir file=lib/ecto/migrator.ex function=log/2 line=707 mfa=Ecto.Migrator.log/2 module=Ecto.Migrator pid=<0.95.0> [info] Migrations already up
06:10:12.906 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/cowboy2_adapter.ex function=start_link/3 line=103 mfa=Phoenix.Endpoint.Cowboy2Adapter.start_link/3 module=Phoenix.Endpoint.Cowboy2Adapter pid=<0.366.0> [info] Running FzHttpWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:13000 (http)
06:10:12.926 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/supervisor.ex function=log_access_url/2 line=425 mfa=Phoenix.Endpoint.Supervisor.log_access_url/2 module=Phoenix.Endpoint.Supervisor pid=<0.331.0> [info] Access FzHttpWeb.Endpoint at https://firezone.coliemore.com.au
06:10:14.020 erl_level=notice domain=otp file=application_controller.erl function=info_exited/3 line=2119 mfa=:application_controller.info_exited/3 module=application_controller pid=<0.44.0> [notice] Application fz_wall exited: FzWall.Application.start(:normal, []) returned an error: shutdown: failed to start child: FzWall.Server
    ** (EXIT) an exception was raised:
        ** (FunctionClauseError) no function clause matching in FzHttp.Types.Int4Range.cast/1
            (fz_http 0.7.34) lib/fz_http/types/int4range.ex:14: FzHttp.Types.Int4Range.cast(nil)
            (fz_wall 0.7.34) lib/fz_wall/cli/live.ex:164: FzWall.CLI.Live.modify_elem/2
            (elixir 1.14.3) lib/enum.ex:980: anonymous fn/3 in Enum.each/2
            (elixir 1.14.3) lib/enum.ex:4307: anonymous fn/3 in Enum.each/2
            (elixir 1.14.3) lib/enum.ex:4751: Enumerable.List.reduce/3
            (elixir 1.14.3) lib/enum.ex:4307: Enum.each/2
            (fz_wall 0.7.34) lib/fz_wall/server.ex:18: FzWall.Server.init/1
            (stdlib 4.2) gen_server.erl:851: :gen_server.init_it/2
{"Kernel pid terminated",application_controller,"{application_start_failure,fz_wall,{{shutdown,{failed_to_start_child,'Elixir.FzWall.Server',{function_clause,[{'Elixir.FzHttp.Types.Int4Range',cast,[nil],[{file,\"lib/fz_http/types/int4range.ex\"},{line,14}]},{'Elixir.FzWall.CLI.Live',modify_elem,2,[{file,\"lib/fz_wall/cli/live.ex\"},{line,164}]},{'Elixir.Enum','-each/2-fun-0-',3,[{file,\"lib/enum.ex\"},{line,980}]},{'Elixir.Enum','-each/2-anonymous-3-',3,[{file,\"lib/enum.ex\"},{line,4307}]},{'Elixir.Enumerable.List',reduce,3,[{file,\"lib/enum.ex\"},{line,4751}]},{'Elixir.Enum',each,2,[{file,\"lib/enum.ex\"},{line,4307}]},{'Elixir.FzWall.Server',init,1,[{file,\"lib/fz_wall/server.ex\"},{line,18}]},{gen_server,init_it,2,[{file,\"gen_server.erl\"},{line,851}]}]}}},{'Elixir.FzWall.Application',start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,fz_wall,{{shutdown,{failed_to_start_child,'Elixir.FzWall.Server',{function_clause,[{'Elixir.FzHttp.Types.Int4Range',cast,[nil],[{file,"lib/fz_http/types/int4range.ex"},{line,14}]},{'Elixir.FzWall.CLI.Live',modify_elem,2,[{file,"lib/fz_wall/cli/live.ex"},{line,164}]},{'Elixir.Enum','-each/2-fun-0-',3,[{file,"lib/enum.ex"},{line,980}]},{'Elixir.Enum','-each/2-anonymous-3-',3,[{file,"lib/enum.ex"},{line,4307}]},{'Elixir.Enumerable.List',reduce,3,[{file,"lib/enum.ex"},{line,4751}]},{'Elixir.Enum',each,2,[{file,"lib/enum.ex"},{line,4307}]},{'Elixir.FzWall.Server',init,1,[{file,"lib/fz_wall/server.ex"},{line,18}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,851}]}]}}},{'Elixir.FzWall.Application',start,[normal,[]]}}})
```

**Comment by laweschan** [comment]
It still has error on 0.7.34, just an easy way to produce:

if I add a deny rule to cover a subnet . said 12.12.12.0/24
apply to all users with all protocol, the firezone container will crash immediately.

it also happen if I all allow rule instead, the problem seems in TCP/UDP port if not defined, eg. using "ALL Protocol"


Crash dump is being written to: erl_crash.dump...done
17:28:43.972 erl_level=info application=ecto_sql domain=elixir file=lib/ecto/migrator.ex function=log/2 line=707 mfa=Ecto.Migrator.log/2 module=Ecto.Migrator pid=<0.95.0> [info] Migrations already up
17:28:44.835 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/cowboy2_adapter.ex function=start_link/3 line=103 mfa=Phoenix.Endpoint.Cowboy2Adapter.start_link/3 module=Phoenix.Endpoint.Cowboy2Adapter pid=<0.375.0> [info] Running FzHttpWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:13000 (http)
17:28:44.839 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/supervisor.ex function=log_access_url/2 line=425 mfa=Phoenix.Endpoint.Supervisor.log_access_url/2 module=Phoenix.Endpoint.Supervisor pid=<0.331.0> [info] Access FzHttpWeb.Endpoint at https://xxxxxxxxxxxxx
17:28:45.001 erl_level=notice domain=otp file=application_controller.erl function=info_exited/3 line=2119 mfa=:application_controller.info_exited/3 module=application_controller pid=<0.44.0> [notice] Application fz_wall exited: FzWall.Application.start(:normal, []) returned an error: shutdown: failed to start child: FzWall.Server
    ** (EXIT) an exception was raised:
        ** (FunctionClauseError) no function clause matching in FzHttp.Types.Int4Range.cast/1
            (fz_http 0.7.34) lib/fz_http/types/int4range.ex:14: FzHttp.Types.Int4Range.cast(nil)
            (fz_wall 0.7.34) lib/fz_wall/cli/live.ex:164: FzWall.CLI.Live.modify_elem/2
            (elixir 1.14.3) lib/enum.ex:980: anonymous fn/3 in Enum.each/2
            (elixir 1.14.3) lib/enum.ex:4307: anonymous fn/3 in Enum.each/2
            (elixir 1.14.3) lib/enum.ex:4751: Enumerable.List.reduce/3
            (elixir 1.14.3) lib/enum.ex:4307: Enum.each/2
            (fz_wall 0.7.34) lib/fz_wall/server.ex:18: FzWall.Server.init/1
            (stdlib 4.2) gen_server.erl:851: :gen_server.init_it/2
{"Kernel pid terminated",application_controller,"{application_start_failure,fz_wall,{{shutdown,{failed_to_start_child,'Elixir.FzWall.Server',{function_clause,[{'Elixir.FzHttp.Types.Int4Range',cast,[nil],[{file,\"lib/fz_http/types/int4range.ex\"},{line,14}]},{'Elixir.FzWall.CLI.Live',modify_elem,2,[{file,\"lib/fz_wall/cli/live.ex\"},{line,164}]},{'Elixir.Enum','-each/2-fun-0-',3,[{file,\"lib/enum.ex\"},{line,980}]},{'Elixir.Enum','-each/2-anonymous-3-',3,[{file,\"lib/enum.ex\"},{line,4307}]},{'Elixir.Enumerable.List',reduce,3,[{file,\"lib/enum.ex\"},{line,4751}]},{'Elixir.Enum',each,2,[{file,\"lib/enum.ex\"},{line,4307}]},{'Elixir.FzWall.Server',init,1,[{file,\"lib/fz_wall/server.ex\"},{line,18}]},{gen_server,init_it,2,[{file,\"gen_server.erl\"},{line,851}]}]}}},{'Elixir.FzWall.Application',start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,fz_wall,{{shutdown,{failed_to_start_child,'Elixir.FzWall.Server',{function_clause,[{'Elixir.FzHttp.Types.Int4Range',cast,[nil],[{file,"lib/fz_http/types/int4range.ex"},{line,14}]},{'Elixir.FzWall.CLI.Live',modify_elem,2,[{file,"lib/fz_wall/cli/live.ex"},{line,164}]},{'Elixir.Enum','-each/2-fun-0-',3,[{file,"lib/enum.ex"},{line,980}]},{'Elixir.Enum','-each/2-anonymous-3-',3,[{file,"lib/enum.ex"},{line,4307}]},{'Elixir.Enumerable.List',reduce,3,[{file,"lib/enum.ex"},{line,4751}]},{'Elixir.Enum',each,2,[{file,"lib/enum.ex"},{line,4307}]},{'Elixir.FzWall.Server',init,1,[{file,"lib/fz_wall/server.ex"},{line,18}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,851}]}]}}},{'Elixir.FzWall.Application',start,[normal,[]]}}})

Crash dump is being written to: erl_crash.dump...done
17:28:47.731 erl_level=info application=ecto_sql domain=elixir file=lib/ecto/migrator.ex function=log/2 line=707 mfa=Ecto.Migrator.log/2 module=Ecto.Migrator pid=<0.95.0> [info] Migrations already up
17:28:48.624 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/cowboy2_adapter.ex function=start_link/3 line=103 mfa=Phoenix.Endpoint.Cowboy2Adapter.start_link/3 module=Phoenix.Endpoint.Cowboy2Adapter pid=<0.375.0> [info] Running FzHttpWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:13000 (http)
17:28:48.628 erl_level=info application=phoenix domain=elixir file=lib/phoenix/endpoint/supervisor.ex function=log_access_url/2 line=425 mfa=Phoenix.Endpoint.Supervisor.log_access_url/2 module=Phoenix.Endpoint.Supervisor pid=<0.331.0> [info] Access FzHttpWeb.Endpoint at https://xxxxxxxxxx

**Comment by wallarug** [comment]
Looks fixed for me now in the latest 0.7.35 release.  Thanks Jamil!

**Comment by Annie602942** [comment]
https://github.com/firezone/firezone/pull/1948
